backend/src/main/java/com/example/backend/use_case/predictions/strategies/CalculateDistanceStrategy.java

```diff
@@ -0,0 +1,143 @@
+                // obj2 has not passed the start of the shape, but there is no previous point in shape. This is an error.
+                throw new IndexOutOfBoundsException("failed because j is out of bounds");
+            }
+        }
+        // Now, distance =
+        // distance from obj1 to shape.get(i) +
+        // distance from shape.get(i) to shape.get(j) +
+        // distance from shape.get(j) to obj2
+
+        if (i > j) {
+            throw new IndexOutOfBoundsException("failed because i=" + i + " and j=" + j);
+        }
+
+        System.out.println("Stop coords is " + shape.get(j).getLatitude() + ", " + shape.get(j).getLongitude());
+        System.out.println("Vehicle coords is " + shape.get(i).getLatitude() + ", " + shape.get(i).getLongitude());
+
+        float startDistance = obj1.distanceTo(shape.get(i));
+        float endDistance = shape.get(j).distanceTo(obj2);
+        float middleDistance = 0;
+        while (i != j) {
+            middleDistance += shape.get(i).distanceTo(shape.get(i + 1));
+            i = (i + 1) % shape.size();
+        }
+
+        System.out.println("Start: " + startDistance + "\nMiddle: " + middleDistance + "\nEnd: " + endDistance);
+        return startDistance + middleDistance + endDistance;
+    }
+
+    private int indexOfClosestPointOnShape(DistanceMeasurable point, ArrayList<Location> shape) {
+        int index = 0;
+        float minDistance = point.distanceTo(shape.get(0));
+        for (int j = 1; j < shape.size(); j++) {
+            float distance = point.distanceTo(shape.get(j));
+            if (distance < minDistance) {
+                minDistance = distance;
+                index = j;
+            }
+        }
+        return index;
+    }
+
+    public ArrayList<Float> execute(StrategyInputData data) {
+        Route route = data.getRoute();
+        String dirTag = data.getDirTag();
+        float avgSpeed = data.getAvgSpeed();
+        RouteDirection direction = route.getRouteDirections().get(dirTag);
+        Stop stop;
+        if (!direction.getStops().contains(data.getStopTag()) || avgSpeed <= 0) {
+            return new ArrayList<>();
+        } else {
+            stop = route.getStops().get(data.getStopTag());
+        }
+        HashSet<Vehicle> vehiclesInDirection = new HashSet<>();
+
+        for (Vehicle vehicle : data.getVehicles()) {
+            if (vehicle.getDirectionTag().equals(dirTag)) {
+                vehiclesInDirection.add(vehicle);
+            }
+        }
+
+        ArrayList<Float> distances = new ArrayList<>();
+        for (Vehicle vehicle : vehiclesInDirection) {
+            System.out.println("Vehicle is at " + vehicle.getLocation().getLatitude() + ", " + vehicle.getLocation().getLongitude());
+            try {
+                float distance = distanceAlongShape(data.getShape(), vehicle, stop);
+                distances.add(distance);
+            } catch (IndexOutOfBoundsException e) {
```

Comment:
Rather than returning the moment you encounter an index out of bounds error, shouldn't you try iterating through all vehicles first, then check at the end if distances is empty?